admin_manual/enterprise/external_storage/s3_swift_as_primary_object_store_configuration.rst

```diff
@@ -1,159 +1,167 @@
+
+.. note:: OpenStack Swift has been deprecated.
+
+When using ``files_primary_s3``, the Amazon S3 bucket needs to be created manually according to the developer documentation, and versioning needs to be enabled.
+
+.. note:: ownCloud GmbH provides consulting for migrations from ``objectstore`` to ``files_primary_s3``.
+
+.. note:: The source of the S3 Object Storage app `is available on gitHub`_.
 
 Implications
 ------------
-This configuration needs to be applied before the first login of any user – including the 
-admin user. Otherwise, you will run into an error condition, because ownCloud cannot find 
-the user's files anymore.
 
-ownCloud in object store mode expects exclusive access to the object store 
-container, because it only stores the binary data for each file. The metadata 
-are kept in the local database for performance reasons.
+- Apply this configuration before the first login of any user – including the admin user; otherwise, ownCloud can no longer find the user's files.
+
+- ownCloud, in "object store" mode, expects exclusive access to the object store container, because it only stores the binary data for each file. While in this mode, ownCloud stores the metadata in the local database for performance reasons.
 
-The current implementation is incompatible with any app that uses 
-direct file I/O and circumvents the ownCloud virtual filesystem. That includes 
-Encryption and Gallery. Gallery stores thumbnails directly in the 
-filesystem, and Encryption causes severe overhead because key files need 
-to be fetched in addition to any requested file.
+- The current implementation is incompatible with any app that uses direct file I/O (input/output) as it circumvents the ownCloud virtual filesystem. Two excellent examples are:
+
+  #. **The Encryption app:** It fetches critical files in addition to any requested file, which results in significant overhead.
+  #. **The Gallery app:** It stores thumbnails directly in the filesystem.
 
 Configuration
 -------------
 
-Look in ``config.sample.php`` for a example configurations. Copy the 
-relevant part to your ``config.php`` file. Any object store needs to implement
-``\\OCP\\Files\\ObjectStore\\IObjectStore`` and can be passed parameters in the
-constructor with the ``arguments`` key:
+Look in ``config.sample.php`` for example configurations.
+Copy the relevant part to your ``config.php`` file.
+Any object store needs to implement ``\\OCP\\Files\\ObjectStore\\IObjectStore``, and can be passed parameters in the constructor with the ``arguments`` key, as in the following example:
 
-::
+.. code-block:: php
 
-    'objectstore' => [
-        'class' => 'Implementation\\Of\\OCP\\Files\\ObjectStore\\IObjectStore',
-        'arguments' => [
-            ...
+    $CONFIG = [
+        'objectstore' => [
+            'class' => 'Implementation\\Of\\OCP\\Files\\ObjectStore\\IObjectStore',
+            'arguments' => [
+                ...
+            ],
         ],
     ],
 
```

Comment:
There is something interresting in comparing Amazon S3, Ceph S3 and Scality S3
The latter two use following additional keys (example below based on ceph)
Maybe a good idea to highlight that there are common keys used in all configurations but Ceph S3 and Scality S3 have additional parameters.
This would ease adoptions to other S3 compatible but non Amazon sources
```
  'use_path_style_endpoint' => true,
  'endpoint' => 'http://ceph:80/',
```

Reply:
@mmattel that is unrelated of ceph/scality - it is related to the configuration of the service providing storage accessible via s3 protocol. 
Refer to the AWS documentation, which provides the ability to use both https://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html

Reply:
I read the text in the link, but I cannot find the match from the two keys to the text.
Maybe we should add some descriptive text for the keys used in general and reference additional related material for further reading. That would make things more clear and reduce complexity and errors. We should not expect all admins beeing highly skilled developers 😄

Reply:
It's about path-style vs host-style access - `'use_path_style_endpoint'` is exactly configuring this.

From the AWS SDK:
```
use_path_style_endpoint: (bool) Set to true to send requests to an S3 path style endpoint by default. Can be enabled or disabled on individual operations by setting '@use_path_style_endpoint\' to true or false. Note: you cannot use it together with an accelerate endpoint.
```
ref: https://docs.aws.amazon.com/aws-sdk-php/v3/api/class-Aws.S3.S3Client.html

So this does not depend on scality/ceph - any s3 "hoster" can provide access to the storage via host-style or path-style. It's a matter of how it is setup - not tied to any Storage per se

Reply:
So, should a note be added, or does the documentation require further clarity/explanation?

Reply:
my pov woul be yes, you see in the threads/discussions many things that need clarification for better understandig.

Reply:
Linking to the AWS docs should be enought from my pov ...

Reply:
@mmattel 

'use_path_style_endpoint'  simply describes, how the bucket is reachable.

path style means: `https://some.random.tld/<bucket>` vs host style: `https://<bucket>.some.random.tld`.
That is essentially the difference and this depends on the provider for the storage.